jina/types/ndarray/__init__.py

```diff
@@ -19,34 +33,244 @@ class BaseNdArray(ProtoTypeMixin):
     :param proto: the protobuf message, when not given then create a new one via :meth:`get_null_proto`
     """
 
-    def __init__(
-        self, proto: Union['PbMessageType', AnyNdArray] = None, *args, **kwargs
-    ):
-        """Set the constructor method."""
-        if proto is not None and isinstance(type(proto), PbMessageType):
-            self._pb_body = proto  # a weak ref/copy
-        else:
-            self._pb_body = self.null_proto()
-            if proto is not None:
-                # casting using the subclass :attr:`value` interface
-                self.value = proto
+    def __init__(self, proto: jina_pb2.NdArrayProto):
+        self._pb_body = proto
 
     @property
-    def null_proto(self) -> 'PbMessageType':
-        """Get the new protobuf representation."""
-        raise NotImplementedError
+    def value(self) -> 'ArrayType':
+        """Return the value in original framework type
 
-    @property
-    def value(self) -> AnyNdArray:
-        """Return the value of the ndarray, in numpy, scipy, tensorflow, pytorch type."""
-        raise NotImplementedError
+        :return: the value of in numpy, scipy, tensorflow, pytorch type."""
+        framework = self._pb_body.cls_name
+
+        if self.is_sparse:
+            if framework == 'scipy':
+                idx, val, shape = self._get_raw_sparse_array()
+                from scipy.sparse import coo_matrix
+
+                x = coo_matrix((val, idx.T), shape=shape)
+                sp_format = self._pb_body.parameters['sparse_format']
+                if sp_format == 'bsr':
+                    return x.tobsr()
+                elif sp_format == 'csc':
+                    return x.tocsc()
+                elif sp_format == 'csr':
+                    return x.tocsr()
+                elif sp_format == 'coo':
+                    return x
+            elif framework == 'tensorflow':
+                idx, val, shape = self._get_raw_sparse_array()
+                from tensorflow import SparseTensor
+
+                return SparseTensor(idx, val, shape)
+            elif framework == 'torch':
+                idx, val, shape = self._get_raw_sparse_array()
+                from torch import sparse_coo_tensor
+
+                return sparse_coo_tensor(idx, val, shape)
+        else:
+            if framework in {'numpy', 'torch', 'paddle', 'tensorflow'}:
+                x = _get_dense_array(self._pb_body.dense)
+                return _to_framework_array(x, framework)
+
+    @staticmethod
+    def unravel(protos: Sequence[jina_pb2.NdArrayProto]) -> 'ArrayType':
+        """Unravel many ndarray-like proto in one-shot, by following the shape
+        and dtype of the first proto.
+
+        :param protos: a list of ndarray protos
+        :return: a framework ndarray
+        """
+        first = NdArray(next(iter(protos)))
```

Comment:
why not simply access 0? no need to create new object

Reply:
because it wont work on DAM, as written here https://github.com/jina-ai/jina/blob/master/jina/types/arrays/mixins/content.py#L25

Reply:
Note that DAM's `self._pb_body` is https://github.com/jina-ai/jina/blob/dd752ff919a9278c8d275710420c00fd551904c4/jina/types/arrays/memmap/__init__.py#L530-L533